# Notebook 2 — Event Classification (Stage A)
**Project:** Not Yet Priced In

This notebook runs the full Stage A NLP pipeline:
1. **A0** — Parse raw 8-K `.txt` files → `clean_text` per filing
2. **A1** — Rule-based event category (SEC Item → broad bucket)
3. **A2** — Structural text features (numeric / forward-looking / financial-symbol density)
4. **A3** — LLM scoring via Claude API (importance 1–5 + grounded key signals)
5. **A4** — Human validation sample export
6. **Merge** — Produce `stage_a_final.csv` for NB3

**Prerequisites:** `filings_raw.csv` and `data/filings_text/*.txt` from NB1-2

---

## Setup

In [1]:
%pip install anthropic pandas numpy scipy tqdm nbformat


[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [4]:
import sys, os, re, json, time, warnings
from pathlib import Path

import pandas as pd
import numpy as np
from scipy.stats import rankdata, spearmanr
from tqdm import tqdm

warnings.filterwarnings("ignore")

# ── Paths ──────────────────────────────────────────────────────────────────
# REPO_ROOT assumes this notebook lives in a notebooks/ subfolder.
# If it's in the repo root, change to: REPO_ROOT = Path(".").resolve()
REPO_ROOT   = Path(".").resolve().parent
DATA_DIR    = REPO_ROOT / "data"
FILINGS_DIR = DATA_DIR / "filings_text"

FILINGS_RAW_CSV   = DATA_DIR / "filings_raw.csv"
PARSE_SUMMARY_CSV = DATA_DIR / "parse_summary.csv"
CLEAN_TEXT_CSV    = DATA_DIR / "parse_summary_clean_text.csv"
STAGE_A1_CSV      = DATA_DIR / "stage_a1_output.csv"
STAGE_A2_CSV      = DATA_DIR / "stage_a2_output.csv"
STAGE_A3_CSV      = DATA_DIR / "stage_a3_output.csv"
STAGE_A_FINAL_CSV = DATA_DIR / "stage_a_final.csv"
A4_SAMPLE_CSV     = DATA_DIR / "a4_human_labels_template.csv"

# Add repo root so parse_filing.py can be imported
sys.path.insert(0, str(REPO_ROOT))

print("Libraries loaded")
print(f"  REPO_ROOT   : {REPO_ROOT}")
print(f"  Filings dir : {FILINGS_DIR}")
print(f"  Exists?     : {FILINGS_DIR.exists()}")

Libraries loaded
  REPO_ROOT   : /Users/dagur/msba/bigdata/not-yet-priced-in
  Filings dir : /Users/dagur/msba/bigdata/not-yet-priced-in/data/filings_text
  Exists?     : True


---
## Stage A0 — Document Parsing

Strips boilerplate from raw 8-K `.txt` files and extracts `clean_text` per filing.

**Input:** `data/filings_text/*.txt`
**Output:** `parse_summary.csv` + `parse_summary_clean_text.csv`

In [5]:
from parse_filing import batch_parse

print("Running A0 — batch parse...")
df_parse = batch_parse(
    filings_dir=str(FILINGS_DIR),
    output_csv=str(PARSE_SUMMARY_CSV),
    verbose=True,
)
print(f"A0 complete — {len(df_parse)} filings parsed")

Running A0 — batch parse...
Parsing 474 filings from /Users/dagur/msba/bigdata/not-yet-priced-in/data/filings_text...
  Processed 50/474
  Processed 100/474
  Processed 150/474
  Processed 200/474
  Processed 250/474
  Processed 300/474
  Processed 350/474
  Processed 400/474
  Processed 450/474

Saved diagnostic summary → /Users/dagur/msba/bigdata/not-yet-priced-in/data/parse_summary.csv
Saved clean text       → /Users/dagur/msba/bigdata/not-yet-priced-in/data/parse_summary_clean_text.csv

══════════════════════════════════════════════════════════════════════
  PARSE QUALITY REPORT
══════════════════════════════════════════════════════════════════════
  Total filings:       474
  Parse success:       459 (96.8%)
  Parse failure:       15 (3.2%)

  ── Parse confidence distribution ──
    0.0-0.3 (critical)          26 filings (5.5%)
    0.3-0.5 (poor)              39 filings (8.2%)
    0.5-0.7 (suspect)            1 filings (0.2%)
    0.7-0.99 (ok-ish)            0 filings (0.0%)
    1

In [6]:
n = len(df_parse)
n_success = df_parse["parse_success"].sum()
n_clean   = (df_parse["parse_confidence"] >= 0.7).sum()

print(f"Parse success : {n_success}/{n}  ({n_success/n:.1%})  — need >95%")
print(f"Confidence>=0.7: {n_clean}/{n}  ({n_clean/n:.1%})  — need >90%")

if n_success / n < 0.95:
    print("WARN: parse success below 95% — investigate before continuing")
else:
    print("OK: parse quality looks good")

Parse success : 459/474  (96.8%)  — need >95%
Confidence>=0.7: 408/474  (86.1%)  — need >90%
OK: parse quality looks good


In [7]:
df = pd.read_csv(CLEAN_TEXT_CSV)
print(f"Clean text CSV: {len(df)} rows")
print(f"Columns: {list(df.columns)}")
df.head(3)

Clean text CSV: 474 rows
Columns: ['filename', 'ticker', 'accession', 'primary_item', 'primary_category', 'clean_text']


,filename,ticker,accession,primary_item,primary_category,clean_text
0,AAPL_0000320193_23_000005.txt,AAPL,0000320193_23_000005,2.02,Results of Operations / Earnings,[Item 2.02 — Results of Operations / Earnings]...
1,AAPL_0000320193_23_000063.txt,AAPL,0000320193_23_000063,2.02,Results of Operations / Earnings,[Item 2.02 — Results of Operations / Earnings]...
2,AAPL_0000320193_23_000075.txt,AAPL,0000320193_23_000075,2.02,Results of Operations / Earnings,[Item 2.02 — Results of Operations / Earnings]...


In [8]:
# Spot-check: CVS Earnings filing (reference example throughout)
cvs = df[(df["ticker"] == "CVS") & (df["primary_item"] == "2.02")].head(1)
if len(cvs):
    r = cvs.iloc[0]
    print(f"Found: {r['filename']}")
    print(f"Clean text length: {len(str(r['clean_text']))} chars")
    print(f"Preview:\n{str(r['clean_text'])[:400]}")
else:
    print("CVS 2.02 not found — check ticker/item columns")

CVS 2.02 not found — check ticker/item columns


---
## Stage A1 — Rule-Based Event Category

Maps each filing's `primary_item` to a broad event bucket. Fully deterministic — no AI.

**Adds column:** `broad_category`

In [9]:
BROAD_CATEGORY_MAP = {
    "2.02": "Earnings",
    "1.01": "Material Agreement",
    "1.02": "Material Agreement",
    "1.03": "Material Agreement",
    "5.02": "Executive Change",
    "5.07": "Voting Results",
    "5.03": "Governance",
    "5.05": "Governance",
    "5.08": "Governance",
    "5.01": "Governance",
    "5.04": "Governance",
    "5.06": "Governance",
    "7.01": "Regulatory",
    "8.01": "Regulatory",
    "2.03": "Financial Obligation",
    "2.04": "Financial Obligation",
    "2.05": "Financial Obligation",
    "2.06": "Financial Obligation",
    "2.01": "Acquisition/Disposition",
    "3.01": "Securities",
    "3.02": "Securities",
    "3.03": "Securities",
    "4.01": "Accounting",
    "4.02": "Accounting",
    "1.04": "Other",
}

def to_broad_category(item):
    if pd.isna(item):
        return "Other"
    return BROAD_CATEGORY_MAP.get(str(item).strip(), "Other")

df["broad_category"] = df["primary_item"].apply(to_broad_category)

print("A1 complete")
print(df["broad_category"].value_counts().to_string())

A1 complete
broad_category
Earnings                   168
Regulatory                 132
Executive Change            72
Voting Results              39
Material Agreement          25
Other                       17
Governance                  14
Financial Obligation         4
Acquisition/Disposition      3


In [10]:
other_pct = (df["broad_category"] == "Other").mean()
print(f"'Other' share: {other_pct:.1%}  — target <10%")

if other_pct > 0.10:
    print("Unknown item codes (add to BROAD_CATEGORY_MAP):")
    print(df[df["broad_category"] == "Other"]["primary_item"].value_counts().head(10).to_string())
else:
    print("OK: Other within acceptable range")

df.to_csv(STAGE_A1_CSV, index=False)
print(f"Saved: {STAGE_A1_CSV}")

'Other' share: 3.6%  — target <10%
OK: Other within acceptable range
Saved: /Users/dagur/msba/bigdata/not-yet-priced-in/data/stage_a1_output.csv


---
## Stage A2 — Structural Text Features

Three rule-based signals from `clean_text`. No AI. Fully transparent.

| Feature | What it measures |
|---|---|
| `numeric_density` | % tokens containing a digit |
| `forward_looking_density` | % tokens that are forward-looking words |
| `financial_symbol_density` | $ and % frequency per 100 words |
| `baseline_importance` | Percentile-rank average of all three (0–1) |

**CVS January targets:** numeric 5–10%, forward-looking 3–6%, financial 1–3%, baseline > 0.5

In [11]:
def numeric_density(text):
    if not text or not isinstance(text, str):
        return 0.0
    words = text.split()
    if not words:
        return 0.0
    return sum(1 for w in words if any(c.isdigit() for c in w)) / len(words) * 100


FORWARD_LOOKING_TERMS = {
    "will", "expect", "expects", "expected", "expecting",
    "anticipate", "anticipates", "anticipated",
    "forecast", "forecasts", "forecasted",
    "project", "projects", "projected", "projecting",
    "estimate", "estimates", "estimated", "estimating",
    "intend", "intends", "intended",
    "plan", "plans", "planned", "planning",
    "believe", "believes", "believed",
    "outlook", "guidance", "target", "targets", "goal", "goals",
    "upcoming", "future", "forward",
    "should", "could", "may", "might", "would",
    "projected", "pending", "approximately",
}
MULTI_WORD_PHRASES = [
    "next quarter", "next year", "going forward",
    "full-year", "full year", "in the future",
]

def forward_looking_density(text):
    if not text or not isinstance(text, str):
        return 0.0
    text_lower = text.lower()
    words = text_lower.split()
    if not words:
        return 0.0
    count = sum(1 for w in words if w.strip(".,;:!?\"'()[]") in FORWARD_LOOKING_TERMS)
    for phrase in MULTI_WORD_PHRASES:
        count += text_lower.count(phrase)
    return count / len(words) * 100


def financial_symbol_density(text):
    if not text or not isinstance(text, str):
        return 0.0
    words = text.split()
    if not words:
        return 0.0
    return (text.count("$") + text.count("%")) / len(words) * 100


print("Feature functions defined")

Feature functions defined


In [12]:
print("Computing A2 features...")
df["numeric_density"]          = df["clean_text"].apply(numeric_density)
df["forward_looking_density"]  = df["clean_text"].apply(forward_looking_density)
df["financial_symbol_density"] = df["clean_text"].apply(financial_symbol_density)


def percentile_rank(series):
    filled = series.fillna(0)
    return rankdata(filled, method="average") / len(filled)


df["nd_pct"]  = percentile_rank(df["numeric_density"])
df["fld_pct"] = percentile_rank(df["forward_looking_density"])
df["fsd_pct"] = percentile_rank(df["financial_symbol_density"])
df["baseline_importance"] = (df["nd_pct"] + df["fld_pct"] + df["fsd_pct"]) / 3

print("A2 complete")
print(df[["numeric_density","forward_looking_density",
          "financial_symbol_density","baseline_importance"]].describe().round(3).to_string())

Computing A2 features...
A2 complete
       numeric_density  forward_looking_density  financial_symbol_density  baseline_importance
count          474.000                  474.000                   474.000              474.000
mean             8.884                    1.121                     0.692                0.501
std              4.855                    1.383                     1.572                0.163
min              0.000                    0.000                     0.000                0.176
25%              5.556                    0.000                     0.000                0.381
50%              8.816                    0.685                     0.000                0.478
75%             11.111                    1.722                     0.562                0.620
max             28.693                    7.432                     8.562                0.967


In [13]:
cvs = df[(df["ticker"] == "CVS") & (df["broad_category"] == "Earnings")].head(1)
if len(cvs):
    r = cvs.iloc[0]
    print("CVS Earnings spot-check:")
    print(f"  numeric_density         : {r['numeric_density']:.2f}%  (target 5-10%)")
    print(f"  forward_looking_density : {r['forward_looking_density']:.2f}%  (target 3-6%)")
    print(f"  financial_symbol_density: {r['financial_symbol_density']:.2f}%  (target 1-3%)")
    print(f"  baseline_importance     : {r['baseline_importance']:.3f}   (target >0.5)")
    if r["baseline_importance"] < 0.5:
        print("  WARN: baseline below 0.5 — check feature computation")
    else:
        print("  OK")

print("\nTop 5 by baseline_importance:")
display(df.nlargest(5, "baseline_importance")[
    ["ticker","broad_category","numeric_density","forward_looking_density","baseline_importance"]
].round(3))

df.to_csv(STAGE_A2_CSV, index=False)
print(f"Saved: {STAGE_A2_CSV}")

CVS Earnings spot-check:
  numeric_density         : 7.84%  (target 5-10%)
  forward_looking_density : 6.78%  (target 3-6%)
  financial_symbol_density: 2.75%  (target 1-3%)
  baseline_importance     : 0.781   (target >0.5)
  OK

Top 5 by baseline_importance:


,ticker,broad_category,numeric_density,forward_looking_density,baseline_importance
111,COST,Regulatory,25.000,5.556,0.967
5,AAPL,Regulatory,14.845,3.918,0.903
235,JNJ,Earnings,15.464,2.062,0.892
48,AXP,Regulatory,16.049,2.469,0.888
116,COST,Executive Change,11.483,4.306,0.864


Saved: /Users/dagur/msba/bigdata/not-yet-priced-in/data/stage_a2_output.csv


---
## Stage A3 — LLM Importance Scoring

Calls the Claude API on each `clean_text` and extracts a structured JSON response:
- `importance_score` — integer 1–5
- `key_signals` — verbatim quotes from the filing (grounding mechanism)
- `reasoning` — why this score, citing the text
- `confidence` — LLM self-assessed confidence

**Cost:** ~few dollars for 474 filings
**Runtime:** ~8–15 min sequentially
**Key check:** `grounding_rate` — fraction of `key_signals` appearing verbatim in source. Target > 0.9

> Set `ANTHROPIC_API_KEY` in your environment before running.

In [14]:


SYSTEM_PROMPT = (
    "You are a financial analyst assistant reading SEC 8-K filings.\n\n"
    "Return valid JSON with exactly these fields:\n"
    "- event_category: one of [Earnings, Material Agreement, Executive Change, "
    "Voting Results, Governance, Regulatory, Financial Obligation, "
    "Acquisition/Disposition, Securities, Accounting, Other]\n"
    "- importance_score: integer 1-5\n"
    "- key_signals: list of 2-5 VERBATIM quotes from the filing. "
    "Each must appear EXACTLY in the source text, character-for-character. "
    "Each quote will be programmatically verified and rejected if it does not match verbatim. "
    "Keep each quote under 30 words.\n"
    "- reasoning: 2-3 sentences citing specific text\n"
    "- confidence: float 0.0-1.0\n\n"
    "RUBRIC:\n"
    "5=Very High: specific financials, guidance changes, major M&A or leadership news\n"
    "4=High: concrete updates with numbers or named parties\n"
    "3=Medium: substantive but routine news\n"
    "2=Low: administrative or procedural items\n"
    "1=Very Low: pure formalities, no business impact\n\n"
    "If no verbatim quotes found, return empty key_signals and confidence below 0.5."
)

USER_PROMPT = (
    "Analyze this 8-K filing. Return only the JSON object, no markdown fences.\n\n"
    "---FILING TEXT---\n"
    "{clean_text}\n"
    "---END FILING TEXT---"
)

print("Prompts defined")

Prompts defined


In [13]:
%pip install openai


[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [15]:
%pip install google-generativeai

  Using cached pyasn1_modules-0.4.2-py3-none-any.whl.metadata (3.5 kB)
  Using cached grpcio-1.80.0-cp311-cp311-macosx_11_0_universal2.whl.metadata (3.8 kB)
INFO: pip is looking at multiple versions of grpcio-status to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of grpcio-status to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 155.1/155.1 kB 3.0 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 11.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 173.3/173.3 kB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 240.6/240.6 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 418.6/418.6 kB 8.8 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.0/15.0 MB 16.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [15]:
%env ANTHROPIC_API_KEY=your_key_here
%env OPENAI_API_KEY=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
#%env GEMINI_API_KEY=AIzaSyCSssm-iZraVK_WtToJp-IPc7vxGg_npQM


env: ANTHROPIC_API_KEY=your_key_here
env: OPENAI_API_KEY=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


In [16]:
import os
import json
import time
import re


import anthropic
from openai import OpenAI
import google.generativeai as genai

def analyze_filing(clean_text, provider="anthropic", model=None, max_chars=15000):
    """
    Unified interface for importance scoring across multiple LLM providers.
    Providers: 'anthropic', 'openai', 'gemini'
    """
    if not clean_text or not isinstance(clean_text, str) or len(clean_text.strip()) < 50:
        return {"llm_event_category": "Other", "importance_score": None, "key_signals": [], 
                "reasoning": "Input too short", "confidence": 0.0, "a3_error": "empty_input"}

    truncated = clean_text[:max_chars]
    prompt = USER_PROMPT.format(clean_text=truncated)
    
    try:
        if provider == "anthropic":
            client = anthropic.Anthropic() # Uses ANTHROPIC_API_KEY
            selected_model = model or "claude-3-5-sonnet-20241022"
            response = client.messages.create(
                model=selected_model,
                max_tokens=1024,
                system=SYSTEM_PROMPT,
                messages=[{"role": "user", "content": prompt}]
            )
            raw_text = response.content[0].text

        elif provider == "openai":
            client = OpenAI() # Uses OPENAI_API_KEY
            selected_model = model or "gpt-4o"
            response = client.chat.completions.create(
                model=selected_model,
                messages=[
                    {"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user", "content": prompt}
                ],
                response_format={"type": "json_object"}
            )
            raw_text = response.choices[0].message.content

        elif provider == "gemini":
            genai.configure(api_key=os.environ["GEMINI_API_KEY"])
            selected_model = model or "gemini-1.5-pro"
            client = genai.GenerativeModel(selected_model, system_instruction=SYSTEM_PROMPT)
            response = client.generate_content(
                prompt,
                generation_config=genai.GenerationConfig(response_mime_type="application/json")
            )
            raw_text = response.text

        # Cleanup & Parse JSON
        raw_text = re.sub(r"```json\n?|```", "", raw_text).strip()
        result = json.loads(raw_text)

        return {
            "llm_event_category": result.get("event_category", "Other"),
            "importance_score":   result.get("importance_score"),
            "key_signals":        result.get("key_signals", []),
            "reasoning":          result.get("reasoning", ""),
            "confidence":         result.get("confidence", 0.0),
            "a3_error":           None,
        }

    except Exception as e:
        return {"llm_event_category": "Other", "importance_score": None, "key_signals": [], 
                "reasoning": str(e), "confidence": 0.0, "a3_error": "api_error"}


In [17]:
def validate_grounding(clean_text, key_signals):
    if not key_signals:
        return {"n_signals": 0, "n_grounded": 0, "grounding_rate": 0.0}
    norm_text = " ".join(str(clean_text).split())
    grounded = [" ".join(str(sig).split()) in norm_text for sig in key_signals]
    return {
        "n_signals":     len(key_signals),
        "n_grounded":    sum(grounded),
        "grounding_rate": round(sum(grounded) / len(grounded), 3) if grounded else 0.0,
    }


In [ ]:
# Single test on CVS Earnings before running full batch (Gemini)
test = df[(df["ticker"] == "CVS") & (df["broad_category"] == "Earnings")].head(1)

if len(test):
    r = test.iloc[0]
    
    # We now pass provider="gemini". 
    # It will automatically use the GEMINI_API_KEY from your environment.
    #result = analyze_filing(r["clean_text"], provider="gemini", model="gemini-2.0-flash")
    result = analyze_filing(r["clean_text"], provider="openai", model="gpt-4o")

    
    grounding = validate_grounding(r["clean_text"], result["key_signals"])
    
    print(f"Provider        : openAi")
    print(f"event_category  : {result['llm_event_category']}  (expected: Earnings)")
    print(f"importance_score: {result['importance_score']}        (expected: 4 or 5)")
    print(f"grounding_rate  : {grounding['grounding_rate']}   (target >0.9)")
    print(f"error           : {result['a3_error']}")
    print(f"reasoning       : {result['reasoning'][:200]}")
    
    print("\nkey_signals:")
    for s in result["key_signals"]:
        print(f"  - {s[:80]}")
        
    if result["importance_score"] and result["importance_score"] >= 4:
        print("\n✅ Test passed — Gemini is working. Safe to run full batch.")
    else:
        print("\n⚠️ WARN: score lower than expected — check if Gemini is capturing the financials correctly.")
else:
    print("❌ CVS Earnings row not found")


Provider        : Gemini
event_category  : Earnings  (expected: Earnings)
importance_score: 5        (expected: 4 or 5)
grounding_rate  : 1.0   (target >0.9)
error           : None
reasoning       : The filing contains substantial financial updates, including EPS guidance reaffirmations and revenue expectations for 2022. Notably, CVS Health anticipates surpassing its revenue guidance, indicating 

key_signals:
  - reaffirmed its estimated full-year 2022 Adjusted earnings per share (“EPS”) guid
  - expects its full-year 2022 Adjusted EPS to be at the high end of the guidance ra
  - expects that it will exceed its full-year 2022 revenue guidance range of $309 bi
  - reaffirm its full-year 2022 adjusted operating income guidance range of $17.5 bi
  - entered into a $2.0 billion fixed dollar accelerated share repurchase transactio

✅ Test passed — Gemini is working. Safe to run full batch.


In [ ]:
print(test) 
r = test.iloc[0]
print('_______')
print(r)
print('_______')
print(r["clean_text"])


                         filename ticker             accession  primary_item  \
121  CVS_0000064803_23_000003.txt    CVS  0000064803_23_000003          2.02   

                     primary_category  \
121  Results of Operations / Earnings   

                                            clean_text broad_category  \
121  [Item 2.02 — Results of Operations / Earnings]...       Earnings   

     numeric_density  forward_looking_density  financial_symbol_density  \
121         7.838983                 6.779661                  2.754237   

      nd_pct   fld_pct   fsd_pct  baseline_importance  
121  0.43038  0.995781  0.915612             0.780591  
_______
filename                                         CVS_0000064803_23_000003.txt
ticker                                                                    CVS
accession                                                0000064803_23_000003
primary_item                                                             2.02
primary_category          

In [26]:
import os
if CHECKPOINT_PATH.exists():
    os.remove(CHECKPOINT_PATH)
    print("Cleared failed checkpoint.")


Cleared failed checkpoint.


In [27]:
CHECKPOINT_PATH = DATA_DIR / "stage_a3_checkpoint.csv"
SLEEP_BETWEEN   = 1.0

# Resume from checkpoint if it exists
if CHECKPOINT_PATH.exists():
    done_df  = pd.read_csv(CHECKPOINT_PATH, index_col="_idx")
    done_idx = set(done_df.index.tolist())
    print(f"Resuming from checkpoint — {len(done_df)} already scored")
else:
    done_df  = pd.DataFrame()
    done_idx = set()

results = []
for idx, row in tqdm(df.iterrows(), total=len(df), desc="A3 LLM scoring"):
    if idx in done_idx:
        continue

    #result    = analyze_filing(row["clean_text"])
    result    = analyze_filing(row["clean_text"], provider="openai", model="gpt-4o")
    grounding = validate_grounding(row["clean_text"], result["key_signals"])

    results.append({
        "_idx":               idx,
        "llm_event_category": result["llm_event_category"],
        "importance_score":   result["importance_score"],
        "key_signals":        json.dumps(result["key_signals"]),
        "reasoning":          result["reasoning"],
        "llm_confidence":     result["confidence"],
        "a3_error":           result["a3_error"],
        "n_signals":          grounding["n_signals"],
        "n_grounded":         grounding["n_grounded"],
        "grounding_rate":     grounding["grounding_rate"],
    })

    if len(results) % 50 == 0:
        ckpt = pd.DataFrame(results).set_index("_idx")
        if not done_df.empty:
            ckpt = pd.concat([done_df, ckpt])
        ckpt.to_csv(CHECKPOINT_PATH)
        tqdm.write(f"  Checkpoint saved: {len(results)} new results")

    time.sleep(SLEEP_BETWEEN)

new_df = pd.DataFrame(results)
if not new_df.empty:
    new_df = new_df.set_index("_idx")

a3_results = pd.concat([done_df, new_df]) if not done_df.empty else new_df

print(f"A3 complete — {len(a3_results)} filings scored")
print(f"Errors        : {a3_results['a3_error'].notna().sum()}")
print(f"Mean grounding: {a3_results['grounding_rate'].mean():.3f}")

A3 LLM scoring:   0%|          | 0/474 [00:00<?, ?it/s]

A3 LLM scoring:  10%|█         | 49/474 [02:16<20:44,  2.93s/it]

  Checkpoint saved: 50 new results


A3 LLM scoring:  21%|██        | 99/474 [04:48<15:24,  2.47s/it]

  Checkpoint saved: 100 new results


A3 LLM scoring:  31%|███▏      | 149/474 [07:11<18:37,  3.44s/it]

  Checkpoint saved: 150 new results


A3 LLM scoring:  42%|████▏     | 199/474 [09:44<20:17,  4.43s/it]

  Checkpoint saved: 200 new results


A3 LLM scoring:  53%|█████▎    | 249/474 [12:19<12:16,  3.27s/it]

  Checkpoint saved: 250 new results


A3 LLM scoring:  63%|██████▎   | 299/474 [15:10<09:58,  3.42s/it]

  Checkpoint saved: 300 new results


A3 LLM scoring:  74%|███████▎  | 349/474 [18:05<07:33,  3.63s/it]

  Checkpoint saved: 350 new results


A3 LLM scoring:  84%|████████▍ | 399/474 [21:03<04:10,  3.34s/it]

  Checkpoint saved: 400 new results


A3 LLM scoring:  95%|█████████▍| 449/474 [24:00<01:32,  3.71s/it]

  Checkpoint saved: 450 new results


A3 LLM scoring: 100%|██████████| 474/474 [25:23<00:00,  3.21s/it]

A3 complete — 474 filings scored
Errors        : 18
Mean grounding: 0.833


In [29]:
# List of columns to clear from the original df if they exist
cols_to_clear = [
    "llm_event_category", "importance_score", "key_signals", "reasoning", 
    "llm_confidence", "n_signals", "n_grounded", "grounding_rate"
]
# Drop overlaps from df first, then join the new results
df = df.drop(columns=cols_to_clear, errors="ignore")
df = df.join(a3_results.drop(columns=["a3_error"], errors="ignore"), how="left")

n_scored = df["importance_score"].notna().sum()
print(f"Scored: {n_scored}/{len(df)}")
print(f"Failed: {df['importance_score'].isna().sum()}")
print(f"Mean grounding_rate: {df['grounding_rate'].mean():.3f}  (target >0.9)")

if df["grounding_rate"].mean() < 0.9:
    print("WARN: grounding below target — consider tightening prompt")

print("\nImportance score distribution:")
print(df["importance_score"].value_counts().sort_index().to_string())

df.to_csv(STAGE_A3_CSV, index=False)
print(f"Saved: {STAGE_A3_CSV}")

Scored: 456/474
Failed: 18
Mean grounding_rate: 0.833  (target >0.9)
WARN: grounding below target — consider tightening prompt

Importance score distribution:
importance_score
1.0     12
2.0     38
3.0    164
4.0    173
5.0     69
Saved: /Users/dagur/msba/bigdata/not-yet-priced-in/data/stage_a3_output.csv


---
## Stage A4 — Human Validation Sample

Exports a stratified sample (~40 filings) for two team members to score independently.

**Steps:**
1. Run the export cell → `a4_human_labels_template.csv`
2. Two people fill `human_A_score` and `human_B_score` independently — do NOT look at `llm_score` first
3. Rename to `a4_human_labels.csv` when complete
4. Run the agreement cell to get the flier number

**Rubric:** 5=Very High, 4=High, 3=Medium, 2=Low, 1=Very Low

In [30]:
SAMPLE_PER_CAT = 5
RANDOM_SEED    = 42

scored_df = df[df["importance_score"].notna()].copy()
parts = []
for cat, grp in scored_df.groupby("broad_category"):
    parts.append(grp.sample(min(SAMPLE_PER_CAT, len(grp)), random_state=RANDOM_SEED))

a4_sample  = pd.concat(parts).sample(frac=1, random_state=RANDOM_SEED)
a4_export  = a4_sample[["filename","ticker","broad_category","primary_item",
                          "clean_text","importance_score"]].copy()
a4_export  = a4_export.rename(columns={"importance_score": "llm_score"})
a4_export["human_A_score"] = ""
a4_export["human_B_score"] = ""

a4_export.to_csv(A4_SAMPLE_CSV, index=False)
print(f"A4 sample exported: {len(a4_export)} filings")
print(f"Saved: {A4_SAMPLE_CSV}")
print()
print("Category breakdown:")
print(a4_export["broad_category"].value_counts().to_string())

A4 sample exported: 37 filings
Saved: /Users/dagur/msba/bigdata/not-yet-priced-in/data/a4_human_labels_template.csv

Category breakdown:
broad_category
Governance                 5
Earnings                   5
Regulatory                 5
Voting Results             5
Material Agreement         5
Executive Change           5
Financial Obligation       4
Acquisition/Disposition    3


In [ ]:
# ── Run after both humans have filled in scores ────────────────────────────
# Rename completed file to a4_human_labels.csv first, then uncomment below.

# from sklearn.metrics import cohen_kappa_score
#
# a4 = pd.read_csv(DATA_DIR / "a4_human_labels.csv")
# a4 = a4.dropna(subset=["human_A_score","human_B_score","llm_score"])
# a4[["human_A_score","human_B_score","llm_score"]] = (
#     a4[["human_A_score","human_B_score","llm_score"]].astype(int)
# )
#
# rho_hh, _ = spearmanr(a4["human_A_score"], a4["human_B_score"])
# kappa_hh  = cohen_kappa_score(
#     a4["human_A_score"], a4["human_B_score"], weights="quadratic"
# )
# a4["human_mean"] = (a4["human_A_score"] + a4["human_B_score"]) / 2
# rho_lh, _  = spearmanr(a4["llm_score"], a4["human_mean"])
#
# print(f"Human-human Spearman rho : {rho_hh:.2f}  (ceiling)")
# print(f"Human-human kappa        : {kappa_hh:.2f}")
# print(f"LLM-vs-human Spearman rho: {rho_lh:.2f}  <- put this on the flier")

print("A4 agreement cells are commented out.")
print("Uncomment and run after a4_human_labels.csv is filled in.")

---
## Final Merge — `stage_a_final.csv`

Joins all A0–A3 columns with filing metadata. This is the NB3 input.

In [31]:
filings_raw = pd.read_csv(FILINGS_RAW_CSV)
filings_raw["accession_norm"] = filings_raw["accession_no"].str.replace("-", "_")

# Extract accession from filename: TICKER_ACCESSION.txt -> ACCESSION
df["accession_norm"] = (
    df["filename"]
    .str.replace(".txt", "", regex=False)
    .str.split("_", n=1)
    .str[1]
)

df_final = df.merge(
    filings_raw[["accession_norm","filed_date","filing_url","company_name"]],
    on="accession_norm",
    how="left",
)
print(f"Rows before merge: {len(df)}   after: {len(df_final)}")
print(f"filed_date coverage: {df_final['filed_date'].notna().sum()}/{len(df_final)}")

Rows before merge: 474   after: 474
filed_date coverage: 474/474


In [32]:
FINAL_COLS = [
    "filename", "ticker", "accession_norm", "filed_date", "company_name",
    "primary_item", "primary_category", "broad_category",
    "numeric_density", "forward_looking_density", "financial_symbol_density",
    "baseline_importance",
    "llm_event_category", "importance_score", "key_signals",
    "reasoning", "llm_confidence", "grounding_rate",
    "clean_text",
]
existing = [c for c in FINAL_COLS if c in df_final.columns]
stage_a_final = df_final[existing].rename(columns={"accession_norm": "accession"})

stage_a_final.to_csv(STAGE_A_FINAL_CSV, index=False)
print(f"stage_a_final.csv saved: {len(stage_a_final)} rows, {len(stage_a_final.columns)} cols")
print(f"Path: {STAGE_A_FINAL_CSV}")

stage_a_final.csv saved: 474 rows, 19 cols
Path: /Users/dagur/msba/bigdata/not-yet-priced-in/data/stage_a_final.csv


In [33]:
print("=" * 55)
print("  STAGE A COMPLETE")
print("=" * 55)
print(f"  Total filings      : {len(stage_a_final)}")
print(f"  With clean_text    : {stage_a_final['clean_text'].notna().sum()}")
print(f"  With importance    : {stage_a_final['importance_score'].notna().sum()}")
print(f"  Mean importance    : {stage_a_final['importance_score'].mean():.2f}")
print(f"  Mean grounding     : {stage_a_final['grounding_rate'].mean():.3f}")
print()
print("  Category breakdown:")
for cat, n in stage_a_final["broad_category"].value_counts().items():
    print(f"    {n:>4}  {cat}")
print()
print(f"  Output : {STAGE_A_FINAL_CSV}")
print(f"  A4     : {A4_SAMPLE_CSV}")
print("  Next   : Run NB3")
print("=" * 55)

  STAGE A COMPLETE
  Total filings      : 474
  With clean_text    : 457
  With importance    : 456
  Mean importance    : 3.55
  Mean grounding     : 0.833

  Category breakdown:
     168  Earnings
     132  Regulatory
      72  Executive Change
      39  Voting Results
      25  Material Agreement
      17  Other
      14  Governance
       4  Financial Obligation
       3  Acquisition/Disposition

  Output : /Users/dagur/msba/bigdata/not-yet-priced-in/data/stage_a_final.csv
  A4     : /Users/dagur/msba/bigdata/not-yet-priced-in/data/a4_human_labels_template.csv
  Next   : Run NB3


In [34]:
print("Top 10 filings by importance score:")
top10 = stage_a_final.nlargest(10, "importance_score")[[
    "ticker", "filed_date", "broad_category",
    "importance_score", "baseline_importance", "grounding_rate",
]].round(3)
display(top10)

Top 10 filings by importance score:


,ticker,filed_date,broad_category,importance_score,baseline_importance,grounding_rate
1,AAPL,2023-05-04,Earnings,5.0,0.463,0.667
9,ABBV,2023-06-29,Executive Change,5.0,0.474,1.000
11,ABBV,2023-11-30,Regulatory,5.0,0.423,1.000
12,ABBV,2023-12-06,Regulatory,5.0,0.421,1.000
13,ABBV,2023-01-06,Earnings,5.0,0.669,1.000
22,ADBE,2023-03-15,Earnings,5.0,0.373,1.000
24,ADBE,2023-06-15,Earnings,5.0,0.373,1.000
28,ADBE,2023-12-18,Material Agreement,5.0,0.442,1.000
29,ADBE,2023-01-19,Material Agreement,5.0,0.505,1.000
52,AXP,2023-06-27,Executive Change,5.0,0.578,1.000
